crates/bindings-csharp/Codegen/Module.cs

```diff
@@ -19,124 +20,197 @@ enum ColumnAttrs : byte
-
-                        if (indexKind.HasFlag(ColumnAttrs.Unique) && !isEquatable)
+                    var fields = GetFields(table)
+                        .Select(f =>
                         {
-                            throw new System.Exception(
-                                $"{f.Type} {f.Name} is not valid for Identity, PrimaryKey or PrimaryKeyAuto as it's not an equatable primitive."
-                            );
-                        }
+                            var indexKind = f.GetAttributes()
+                                .Where(a =>
+                                    a.AttributeClass?.ToString() == "SpacetimeDB.ColumnAttribute"
+                                )
+                                .Select(a => (ColumnAttrs)a.ConstructorArguments[0].Value!)
+                                .SingleOrDefault();
 
-                        return (
-                            f.Name,
-                            Type: SymbolToName(f.Type),
-                            TypeInfo: GetTypeInfo(f.Type),
-                            IndexKind: indexKind,
-                            IsEquatable: isEquatable
-                        );
-                    })
-                    .ToArray();
-
-                return new
-                {
-                    Scope = new Scope(tableSyntax),
-                    table.Name,
-                    FullName = SymbolToName(table),
-                    Fields = fields,
-                    Public = context
+                            return new ColumnDeclaration(f.Name, f.Type, indexKind);
+                        });
+
+                    var isPublic = context
                         .Attributes.SelectMany(attr => attr.NamedArguments)
-                        .Any(pair => pair.Key == "Public" && pair.Value.Value is true)
-                };
-            }
-        );
+                        .Any(pair => pair.Key == "Public" && pair.Value.Value is true);
+
+                    return new TableDeclaration(tableSyntax, table, fields, isPublic);
+                }
+            )
+            .WithTrackingName("SpacetimeDB.Table.Parse");
 
         tables
             .Select(
                 (t, ct) =>
                 {
-                    var autoIncFields = t.Fields.Where(f =>
-                        f.IndexKind.HasFlag(ColumnAttrs.AutoInc)
-                    )
+                    var autoIncFields = t.Fields.Where(f => f.Attrs.HasFlag(ColumnAttrs.AutoInc))
                         .Select(f => f.Name);
 
                     var extensions =
                         $@"
-                            private static readonly Lazy<SpacetimeDB.RawBindings.TableId> tableId = new (() => SpacetimeDB.Runtime.GetTableId(nameof({t.Name})));
+                            private static readonly Lazy<SpacetimeDB.RawBindings.TableId> tableId = new (() => SpacetimeDB.Runtime.GetTableId(nameof({t.ShortName})));
```

Comment:
It might be worth having a comment somewhere documenting the intended namespace structure of the generated code

Reply:
Maybe when it's finalized just a bit more... For now the test snapshots should work as sort of a document.